modules/packages/Python.chpl

```diff
@@ -0,0 +1,1417 @@
+      this.checkException();
+      Py_DECREF(path);
+      Py_DECREF(sys);
+    }
+    @chpldoc.nodoc
+    proc deinit() {
+      Py_Finalize();
+    }
+    /*
+      Run a string of python code.
+
+      This function will just run the code, it cannot be passed arguments or
+      return values.
+    */
+    proc run(code: string) throws {
+      PyRun_SimpleString(code.c_str());
+      this.checkException();
+    }
+
+    /*
+      Register a custom :type:`TypeConverter` with the interpreter. This allows
+      arbitrary Chapel types to be serialized/deserialized to and from Python
+      objects.
+
+      The registered converter will take precedence over the default converters,
+      allowing users to override the default behavior.
+    */
+    proc registerConverter(in cvt: owned TypeConverter) {
+      converters.pushBack(cvt);
+    }
+
+    /*
+      Compile a lambda function like 'lambda x,: x + 1'
+
+      Note: this only works with lambdas that a tuple of arguments, like 'x,' or '*args'
+    */
+    @chpldoc.nodoc
+    proc compileLambda(l: string): owned PyObject throws {
+      var globals = chpl_PyEval_GetFrameGlobals();
+      // create a globals if it doesn't exist
+      if !globals {
+        globals = Py_BuildValue("{}"); // TODO: need to decrement this?
+      }
+      // this is the equivalent of `eval(compile(l, '<string>', 'eval'))`
+      // we can also do `Py_CompileString` -> `PyFunction_New`?
+      var code = PyRun_String(l.c_str(), Py_eval_input, globals, nil);
+      this.checkException();
+      var obj = new PyObject(this, code);
+      return obj;
+    }
+
+    /*
+      Query to see if an Python exception has occurred. If there is an
+      exception, the Python exception will be thrown as a :type:`Exception`.
+
+      .. note::
+
+         This method should be called after any Python API call that may fail.
+         The wrapping types in this module will call this method automatically,
+         most users should not need to call this method directly.
+    */
+    inline proc checkException() throws {
+      var ptype, pvalue, ptraceback: c_ptr(void);
+      PyErr_Fetch(c_ptrTo(ptype), c_ptrTo(pvalue), c_ptrTo(ptraceback));
+
+      if pvalue {
+        var pvalue_str = PyObject_Str(pvalue);
+        var value_str = this.fromPython(string, pvalue_str);
+        Py_DECREF(pvalue_str);
+        throw new PythonException(value_str);
+      }
+    }
+
+    /*
+      Convert a Chapel value to a python object. This clones the Chapel value.
+
+      .. note::
+
+         Most users should not need to call this directly, except when writing
+         a :type:`TypeConverter`.
+    */
+    proc toPython(const val: ?t): c_ptr(void) throws {
+      for converter in this.converters {
+        if converter.handlesType(t) {
+          return converter.toPython(this, t, val);
+        }
+      }
+      // if no converter is found, use the defaults
+      if isIntType(t) {
```

Comment:
can this be a `case` statement?

Reply:
It cannot, since many of the cases rely on more than just simple equality between types.